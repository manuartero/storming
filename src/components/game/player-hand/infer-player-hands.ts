const players: Player[] = ["player", "enemy1", "enemy2", "enemy3"];
const actionCardIds = [
  "build_A",
  "move_A",
  "move_B",
  "recruit_A",
  "diplo_A",
] as const;

function getActionCardType(actionId: string): ActionCardType {
  return actionId.split("_")[0] as ActionCardType;
}

function getPlayedActionCards(timeline: Timeline): ActionCard[] {
  const playedCards = [...timeline.next, ...timeline.future];
  if (timeline.current) {
    playedCards.unshift(timeline.current); // insert timeline.current >> [X...]
  }
  return playedCards.filter(
    (playedCard) => playedCard.cardType === "actionCard"
  ) as ActionCard[];
}

function defineCardStatus({
  cardId,
  selectedCard,
  playedActionCards,
}: {
  playedActionCards: ActionCard[];
  cardId: string;
  selectedCard?: ActionCard;
}): PlayerHandCardStatus {
  if (selectedCard && selectedCard.cardId === cardId) {
    return "selected";
  }
  const hasBeenPlayed = playedActionCards.some(
    (playedCard) => playedCard.cardId === cardId
  );
  return hasBeenPlayed ? "played" : "available";
}

/**
 * {
 *   "player": [ { card, status }, { card, status }, { card, status }... ]
 *   "enemy1": [ ... ]
 *   "enemy2": [ ... ]
 *   "enemy3": [ ... ]
 * }
 */
export function inferPlayerHandsFromGameContext(
  timeline: Timeline,
  selectedCard?: ActionCard
): Record<Player, PlayerHand> {
  const playedActionCards = getPlayedActionCards(timeline);

  return players.reduce((acc, player) => {
    const playerHand: PlayerHand = actionCardIds.map((actionId) => {
      const cardId = `${player}_${actionId}`;
      const card: ActionCard = {
        cardType: "actionCard",
        action: getActionCardType(actionId),
        cardId,
        owner: player,
      };
      const status = defineCardStatus({
        cardId,
        selectedCard,
        playedActionCards,
      });
      return { card, status };
    });

    return {
      ...acc,
      [player]: playerHand,
    };
  }, {} as Record<Player, PlayerHand>);
}
