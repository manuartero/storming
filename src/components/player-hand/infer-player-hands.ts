import { isActionCard } from "models/new-card";
import { PLAYER_CARDS } from "models/player-cards";

function getPlayedActionCards({
  activeCard,
  next,
  future,
}: {
  activeCard: Card | undefined;
  next: TimelineCard[];
  future: TimelineCard[];
}): ActionCard[] {
  const playedCards = [...next, ...future];
  if (activeCard && isActionCard(activeCard)) {
    // insert timeline.current >> [X...]
    playedCards.unshift({ card: activeCard, commited: true });
  }
  return playedCards
    .filter((playedCard) => isActionCard(playedCard.card))
    .map((playedCard) => playedCard.card as ActionCard);
}

function defineCardStatus({
  cardId,
  nextCard,
  playedActionCards,
}: {
  cardId: CardId;
  nextCard?: ActionCard;
  playedActionCards: ActionCard[];
}): PlayerHandCardStatus {
  if (nextCard && nextCard.cardId === cardId) {
    return "selected";
  }
  const hasBeenPlayed = playedActionCards.some(
    (playedCard) => playedCard.cardId === cardId
  );
  return hasBeenPlayed ? "played" : "available";
}

/**
 * {
 *   "player": [ { card, status }, { card, status }, { card, status }... ]
 *   "enemy1": [ ... ]
 *   "enemy2": [ ... ]
 *   "enemy3": [ ... ]
 * }
 */
export function inferPlayerHandsFromGameContext(
  playedCards: {
    activeCard: Card | undefined;
    next: TimelineCard[];
    future: TimelineCard[];
  },
  nextCard?: ActionCard
) {
  const playedActionCards = getPlayedActionCards(playedCards);
  const players = Object.keys(PLAYER_CARDS) as PlayerType[];
  const playerHands = players.reduce((acc, player) => {
    const playerHand = PLAYER_CARDS[player].map((card) => {
      const status = defineCardStatus({
        nextCard,
        cardId: card.cardId,
        playedActionCards,
      });
      return { card, status };
    });
    return {
      ...acc,
      [player]: playerHand,
    };
  }, {} as Record<PlayerType, PlayerHand>);
  return playerHands;
}
