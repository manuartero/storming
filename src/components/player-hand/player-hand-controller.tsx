import { useGameContext } from "contexts";
import { useEffect, useState } from "react";
import PlayerHand from "./player-hand";

const players: Player[] = ["player", "enemy1", "enemy2", "enemy3"];
const actionCards: ActionCardType[] = [
  "build",
  "diplo",
  "move",
  "move",
  "recruit",
];

function getPlayedActionCards(timeline: Timeline): ActionCard[] {
  const playedCards = [...timeline.next, ...timeline.future];
  if (timeline.current) {
    playedCards.unshift(timeline.current);
  }
  return playedCards.filter(
    (playedCard) => playedCard.cardType === "actionCard"
  ) as ActionCard[];
}

function defineCardStatus(
  playedActionCards: ActionCard[],
  player: Player,
  action: ActionCardType,
  selectedCard?: ActionCard
): PlayerHandCardStatus {
  if (selectedCard && selectedCard.action === action) {
    return "selected";
  }
  const hasBeenPlayed = playedActionCards.some(
    (playedCard) => playedCard.owner === player && playedCard.action === action
  );
  return hasBeenPlayed ? "played" : "available";
}

/**
 * {
 *   "player": [ { cardType: "build", status: "available" } ]
 *   "enemy1": [ ... ]
 *   "enemy2": [ ... ]
 *   "enemy3": [ ... ]
 * }
 */
function inferPlayerHandsFromGameContext(
  timeline: Timeline,
  selectedCard?: ActionCard
): Record<Player, PlayerHand> {
  const playedActionCards = getPlayedActionCards(timeline);

  return players.reduce((acc, player) => {
    const playerHand = actionCards.map((action) => {
      const status = defineCardStatus(
        playedActionCards,
        player,
        action,
        selectedCard
      );
      return {
        action,
        status,
      };
    });

    return {
      ...acc,
      [player]: playerHand,
    };
  }, {} as Record<Player, PlayerHand>);
}

function PlayerHandController(): JSX.Element {
  const gameContext = useGameContext();
  const [nextCard, setNextCard] = useState<ActionCard | undefined>(undefined);

  // FIXME can I change the approach: new player -> no selected card
  useEffect(() => {
    setNextCard(undefined);
  }, [gameContext.activePlayer]);

  /* infered state */
  const playerHands = inferPlayerHandsFromGameContext(
    gameContext.timeline,
    nextCard
  );
  const activePlayerHand = gameContext.activePlayer
    ? playerHands[gameContext.activePlayer]
    : [];

  const onClick = (actionCard: ActionCard) => {
    if (!gameContext.activePlayer) {
      console.warn(
        `Inconsistent state: PlayerHandController trying to handle click on ${actionCard} while no activePlayer`
      );
      return;
    }
    if (nextCard) {
      gameContext.plan({
        nextCard,
        futureCard: actionCard,
        player: gameContext.activePlayer,
      });
    } else {
      setNextCard(actionCard);
    }
  };

  return (
    <PlayerHand
      isActive={gameContext.phase === "planification"}
      player={gameContext.activePlayer}
      cards={activePlayerHand}
      onClick={onClick}
    />
  );
}

export default PlayerHandController;
