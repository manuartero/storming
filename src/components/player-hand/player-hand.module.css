.playerHand {
  /** main grid layout */
  grid-area: player-hand;

  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  margin: 0 8vh;
  transition: transform 0.2s ease-in-out;
}

.playerHand.expansible:hover {
  transform: translateY(-40%);
}
