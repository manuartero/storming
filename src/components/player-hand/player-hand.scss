.player-hand {
  display: flex;
  flex-direction: row;
  justify-content: space-evenly;
  align-items: center;

  transition: transform 0.2s ease-in-out;

  &--expansible:hover {
    transform: translateY(-51%);
  }
}
