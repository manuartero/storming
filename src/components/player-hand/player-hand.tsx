import c from "classnames";
import { Card } from "components/cards/card";

import styles from "./player-hand.module.css";

type Props = {
  cards: PlayerHand;
  isActive?: boolean;
  player?: PlayerType;
  onClick: (cardId: CardId) => void;
};

export function PlayerHand({
  cards,
  isActive = false,
  player = undefined,
  onClick,
}: Props) {
  const considerOnClickIfActive = (cardId: CardId) => {
    isActive && onClick(cardId);
  };

  return (
    <section
      role="region"
      className={c(styles.playerHand, isActive && styles.expansible)}
      aria-label="player hand"
    >
      {player &&
        cards.map(({ card, status }) => {
          return (
            <Card
              key={card.cardId}
              status={status}
              card={card}
              onClick={() => {
                considerOnClickIfActive(card.cardId);
              }}
            ></Card>
          );
        })}
    </section>
  );
}
