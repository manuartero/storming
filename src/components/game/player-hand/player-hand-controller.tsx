import { useGameContext } from "contexts";
import { useEffect, useState } from "react";
import { warnInconsistentState } from "utils/console";
import { inferPlayerHandsFromGameContext } from "./infer-player-hands";
import { PlayerHand } from "./player-hand";

/**
 * Renders depends on:
 *  - `useGameContext()`
 *
 * Defines visual player hand from GameContext
 *
 */
function PlayerHandController(): JSX.Element {
  const gameContext = useGameContext();
  const [nextCard, setNextCard] = useState<ActionCard>();

  // FIXME can I change the approach: new player -> no selected card
  useEffect(() => {
    setNextCard(undefined);
  }, [gameContext.activePlayer]);

  /* infered state */
  const playerCards = inferPlayerHandsFromGameContext(
    gameContext.timeline,
    nextCard
  );
  const activePlayerHand = gameContext.activePlayer
    ? playerCards[gameContext.activePlayer]
    : [];

  const handlePlayerCardClick = (cardId: CardId) => {
    if (!gameContext.activePlayer) {
      return warnInconsistentState(
        `trying to handle click on ${cardId} while no activePlayer`
      );
    }
    const playerCard = activePlayerHand.find((e) => e.card.cardId === cardId);
    if (!playerCard) {
      return warnInconsistentState(
        `trying to handle click on ${cardId}, can't find this card on player's hand`,
        { activePlayerHand }
      );
    }

    if (playerCard.status !== "available") {
      return;
    }

    if (nextCard) {
      gameContext.plan({
        nextCard,
        futureCard: playerCard.card as ActionCard, // casting? XXX
        player: gameContext.activePlayer,
      });
    } else {
      setNextCard(playerCard.card as ActionCard); // casting? XXX
    }
  };

  return (
    <PlayerHand
      isActive={gameContext.phase === "planification"}
      player={gameContext.activePlayer}
      cards={activePlayerHand}
      onClick={handlePlayerCardClick}
    />
  );
}

export default PlayerHandController;
