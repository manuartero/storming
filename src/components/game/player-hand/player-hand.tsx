import { Card } from "components/game/cards/card";
import { logRender } from "utils/console";

import "./player-hand.scss";

interface Props {
  cards: PlayerHand;
  isActive?: boolean;
  player?: PlayerType;
  onClick: (cardId: string) => void;
}

function PlayerHand({
  cards,
  isActive = false,
  player = undefined,
  onClick,
}: Props): JSX.Element {
  logRender("PlayerHand");

  const considerOnClickIfActive = (cardId: string) => {
    isActive && onClick(cardId);
  };

  return (
    <div className="player-hand">
      {player &&
        cards.map(({ card, status }) => {
          return (
            <Card
              key={card.cardId}
              status={status}
              card={card}
              onClick={() => considerOnClickIfActive(card.cardId)}
            ></Card>
          );
        })}
    </div>
  );
}

export default PlayerHand;
