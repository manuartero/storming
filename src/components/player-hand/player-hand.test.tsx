import { render, screen } from "@testing-library/react";
import { NewCard } from "models/new-card";
import { PlayerHand } from "./player-hand";

const cards: PlayerHand = [
  {
    card: NewCard("move", "player"),
    status: "available",
  },
  {
    card: NewCard("move", "player"),
    status: "available",
  },
  {
    card: NewCard("build", "player"),
    status: "available",
  },
];

describe("<PlayerHand />", () => {
  test("render: region role", () => {
    render(<PlayerHand cards={cards} player="player" onClick={jest.fn()} />);

    const playerHand = screen.getByRole("region", { name: "player hand" });
    expect(playerHand).toMatchSnapshot();
  });
});
