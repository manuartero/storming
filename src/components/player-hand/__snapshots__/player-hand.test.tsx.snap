// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<PlayerHand /> render: region role 1`] = `
<section
  aria-label="player hand"
  class="playerHand"
  role="region"
>
  <div
    aria-disabled="false"
    aria-label="card player_move_1"
    aria-pressed="false"
    class="card available actionCard clickable"
  >
    <div
      class="actionCardIcon"
    />
    <div
      class="actionCardTitle title"
    >
      move
    </div>
    <div
      class="actionCardText paragraph"
    >
      Move up to two different Units
    </div>
    <div
      class="actionCardWaterMark"
    />
  </div>
  <div
    aria-disabled="false"
    aria-label="card player_move_2"
    aria-pressed="false"
    class="card available actionCard clickable"
  >
    <div
      class="actionCardIcon"
    />
    <div
      class="actionCardTitle title"
    >
      move
    </div>
    <div
      class="actionCardText paragraph"
    >
      Move up to two different Units
    </div>
    <div
      class="actionCardWaterMark"
    />
  </div>
  <div
    aria-disabled="false"
    aria-label="card player_build_1"
    aria-pressed="false"
    class="card available actionCard clickable"
  >
    <div
      class="actionCardIcon"
    />
    <div
      class="actionCardTitle title"
    >
      build
    </div>
    <div
      class="actionCardText paragraph"
    >
      2x:
 Found a Settlement
 Improve a City
Build Walls
    </div>
    <div
      class="actionCardWaterMark"
    />
  </div>
</section>
`;
